File: config/config_manager.py
# config/config_manager.py

import os
import json
import logging
from pathlib import Path
from typing import Dict

from core.utilities import Colors

class ConfigManager:
    def __init__(self, config_path: str = "server_config.json"):
        self.config_path = config_path
        self._config = self.load_config()

    @property
    def config(self) -> Dict:
        return self._config

    def load_config(self) -> Dict:
        """Load configuration from JSON file or create default if not exists."""
        default_config = {
            # Server Directory Settings
            "server_dir": str(Path.home() / "EnshroudedServer"),
            "steamcmd_dir": str(Path.home() / "steamcmd"),
            "backup_dir": str(Path.home() / "EnshroudedBackups"),
            "app_id": "2278520",
            
            # Backup Settings
            "max_backups": 5,
            
            # Server Management
            "auto_restart": True,
            "restart_interval": 24,
            "restart_warning_minutes": 5,
            
            # Game Server Settings
            "server_name": "My Enshrouded Server",
            "max_players": 16,
            "game_port": 15636,
            "query_port": 15637,
            
            # User Groups
            "admin_password": "Admin6cnHwt-D",
            "friend_password": "Friend8p[O-PxS",
            "guest_password": "Guest8AUvMEaA",
            
            # Game Rules
            "game_difficulty": "normal",
            "pvp_enabled": False,
            "player_damage_multiplier": 1.0,
            "enemy_damage_multiplier": 1.0,
            "enemy_health_multiplier": 1.0,
            "day_length_minutes": 30,
            "night_length_minutes": 12,
            
            # Performance Settings
            "max_fps": 60,
            "tick_rate": 64,
            
            # Advanced Settings
            "startup_params": "",
            "allow_commands": True,
            "log_level": "info"
        }

        try:
            if os.path.exists(self.config_path):
                with open(self.config_path, 'r') as f:
                    return {**default_config, **json.load(f)}
            else:
                with open(self.config_path, 'w') as f:
                    json.dump(default_config, f, indent=4)
                return default_config
        except Exception as e:
            logging.error(f"Error loading config: {e}. Using defaults.")
            return default_config

    def update_server_json(self, server_dir: str) -> bool:
        """Update the server's enshrouded_server.json configuration file."""
        try:
            server_json_path = Path(server_dir) / "enshrouded_server.json"
            
            server_json = {
                "name": self.config["server_name"],
                "saveDirectory": "./savegame",
                "logDirectory": "./logs",
                "ip": "0.0.0.0",
                "queryPort": self.config["query_port"],
                "slotCount": self.config["max_players"],
                "gameSettingsPreset": "Default",
                "gameSettings": {
                    "playerHealthFactor": 1,
                    "playerManaFactor": 1,
                    "playerStaminaFactor": 1,
                    "playerBodyHeatFactor": 1,
                    "enableDurability": True,
                    "enableStarvingDebuff": False,
                    "foodBuffDurationFactor": 1,
                    "fromHungerToStarving": 600000000000,
                    "shroudTimeFactor": 1,
                    "tombstoneMode": "AddBackpackMaterials",
                    "enableGliderTurbulences": True,
                    "weatherFrequency": "Normal",
                    "miningDamageFactor": 1,
                    "plantGrowthSpeedFactor": 1,
                    "resourceDropStackAmountFactor": 1,
                    "factoryProductionSpeedFactor": 1,
                    "perkUpgradeRecyclingFactor": 0.5,
                    "perkCostFactor": 1,
                    "experienceCombatFactor": 1,
                    "experienceMiningFactor": 1,
                    "experienceExplorationQuestsFactor": 1,
                    "randomSpawnerAmount": "Normal",
                    "aggroPoolAmount": "Normal",
                    "enemyDamageFactor": self.config["enemy_damage_multiplier"],
                    "enemyHealthFactor": self.config["enemy_health_multiplier"],
                    "enemyStaminaFactor": 1,
                    "enemyPerceptionRangeFactor": 1,
                    "bossDamageFactor": 1,
                    "bossHealthFactor": 1,
                    "threatBonus": 1,
                    "pacifyAllEnemies": False,
                    "tamingStartleRepercussion": "LoseSomeProgress",
                    "dayTimeDuration": self.config["day_length_minutes"] * 60000000000,
                    "nightTimeDuration": self.config["night_length_minutes"] * 60000000000
                },
                "userGroups": [
                    {
                        "name": "Admin",
                        "password": self.config.get("admin_password", "Admin6cnHwt-D"),
                        "canKickBan": True,
                        "canAccessInventories": True,
                        "canEditBase": True,
                        "canExtendBase": True,
                        "reservedSlots": 0
                    },
                    {
                        "name": "Friend",
                        "password": self.config.get("friend_password", "Friend8p[O-PxS"),
                        "canKickBan": False,
                        "canAccessInventories": True,
                        "canEditBase": True,
                        "canExtendBase": False,
                        "reservedSlots": 0
                    },
                    {
                        "name": "Guest",
                        "password": self.config.get("guest_password", "Guest8AUvMEaA"),
                        "canKickBan": False,
                        "canAccessInventories": False,
                        "canEditBase": False,
                        "canExtendBase": False,
                        "reservedSlots": 0
                    }
                ]
            }

            if server_json_path.exists():
                try:
                    with open(server_json_path, 'r') as f:
                        existing_config = json.load(f)
                        if "userGroups" in existing_config:
                            server_json["userGroups"] = existing_config["userGroups"]
                        existing_config.update(server_json)
                        server_json = existing_config
                except json.JSONDecodeError:
                    logging.warning("Could not parse existing server JSON, creating new one")

            with open(server_json_path, 'w') as f:
                json.dump(server_json, f, indent=4)
            
            logging.info(f"Updated server configuration at {server_json_path}")
            return True
        except Exception as e:
            logging.error(f"Failed to update server JSON: {e}")
            return False

    def save_config(self) -> bool:
        """Save the current configuration to file."""
        try:
            with open(self.config_path, 'w') as f:
                json.dump(self.config, f, indent=4)
            return True
        except Exception as e:
            logging.error(f"Error saving config: {e}")
            return False

    def update_config(self, key: str, value: any) -> bool:
        """Update a specific configuration value."""
        try:
            self._config[key] = value
            return self.save_config()
        except Exception as e:
            logging.error(f"Error updating config: {e}")
            return False

File: Config/ConfigManager.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using EnshroudedServerManager.Core;
using EnshroudedServerManager.Models;

namespace EnshroudedServerManager.Config;

public class ConfigManager
{
    // Our app config (server_config.json) uses snake_case
    private static readonly JsonSerializerOptions AppJsonOptions = new()
    {
        PropertyNamingPolicy    = JsonNamingPolicy.SnakeCaseLower,
        WriteIndented           = true,
        ReadCommentHandling     = JsonCommentHandling.Skip,
        AllowTrailingCommas     = true
    };

    // The game's enshrouded_server.json uses camelCase
    private static readonly JsonSerializerOptions GameJsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented        = true
    };

    private readonly string _configPath;
    public ServerConfig Config { get; private set; }

    public ConfigManager(string configPath = "server_config.json")
    {
        _configPath = configPath;
        Config = Load();
    }

    private ServerConfig Load()
    {
        try
        {
            if (File.Exists(_configPath))
            {
                var json   = File.ReadAllText(_configPath);
                var loaded = JsonSerializer.Deserialize<ServerConfig>(json, AppJsonOptions);
                if (loaded != null)
                {
                    AppLogger.Info($"Configuration loaded from {_configPath}");
                    return loaded;
                }
            }
        }
        catch (Exception ex)
        {
            AppLogger.Error($"Error loading config: {ex.Message}. Using defaults.");
        }

        var defaults = new ServerConfig();
        Save(defaults);
        AppLogger.Info("Created default configuration.");
        return defaults;
    }

    public bool Save() => Save(Config);

    private bool Save(ServerConfig config)
    {
        try
        {
            File.WriteAllText(_configPath, JsonSerializer.Serialize(config, AppJsonOptions));
            return true;
        }
        catch (Exception ex)
        {
            AppLogger.Error($"Error saving config: {ex.Message}");
            return false;
        }
    }

    // ── Write enshrouded_server.json ─────────────────────────────────────────
    public bool UpdateServerJson(string serverDir)
    {
        try
        {
            var serverJsonPath = Path.Combine(serverDir, "enshrouded_server.json");
            var gs = Config.GameSettings;

            // The server exits immediately on startup if any user group has an
            // empty password — generate one for each blank group and persist it
            // so the user can read it back in Settings → User Groups.
            bool generated = false;
            foreach (var group in Config.UserGroups)
            {
                if (string.IsNullOrWhiteSpace(group.Password))
                {
                    group.Password = GeneratePassword(group.Name);
                    generated = true;
                    AppLogger.Warning($"User group '{group.Name}' had an empty password — generated '{group.Password}'. " +
                                      "(The server refuses to start when a group password is blank.)");
                }
            }
            if (generated)
                Save();

            // Build the gameSettings block — convert minute values to nanoseconds
            var gameSettingsObj = new
            {
                playerHealthFactor              = gs.PlayerHealthFactor,
                playerManaFactor               = gs.PlayerManaFactor,
                playerStaminaFactor            = gs.PlayerStaminaFactor,
                playerBodyHeatFactor           = gs.PlayerBodyHeatFactor,
                playerDivingTimeFactor         = gs.PlayerDivingTimeFactor,
                enableDurability               = gs.EnableDurability,
                enableStarvingDebuff           = gs.EnableStarvingDebuff,
                foodBuffDurationFactor         = gs.FoodBuffDurationFactor,
                fromHungerToStarving           = (long)(gs.HungerToStarvingMinutes * 60_000_000_000L),
                shroudTimeFactor               = gs.ShroudTimeFactor,
                tombstoneMode                  = gs.TombstoneMode,
                enableGliderTurbulences        = gs.EnableGliderTurbulences,
                weatherFrequency               = gs.WeatherFrequency,
                fishingDifficulty              = gs.FishingDifficulty,
                miningDamageFactor             = gs.MiningDamageFactor,
                plantGrowthSpeedFactor         = gs.PlantGrowthSpeedFactor,
                resourceDropStackAmountFactor  = gs.ResourceDropStackAmountFactor,
                factoryProductionSpeedFactor   = gs.FactoryProductionSpeedFactor,
                perkUpgradeRecyclingFactor     = gs.PerkUpgradeRecyclingFactor,
                perkCostFactor                 = gs.PerkCostFactor,
                experienceCombatFactor         = gs.ExperienceCombatFactor,
                experienceMiningFactor         = gs.ExperienceMiningFactor,
                experienceExplorationQuestsFactor = gs.ExperienceExplorationQuestsFactor,
                randomSpawnerAmount            = gs.RandomSpawnerAmount,
                aggroPoolAmount                = gs.AggroPoolAmount,
                enemyDamageFactor              = gs.EnemyDamageFactor,
                enemyHealthFactor              = gs.EnemyHealthFactor,
                enemyStaminaFactor             = gs.EnemyStaminaFactor,
                enemyPerceptionRangeFactor     = gs.EnemyPerceptionRangeFactor,
                bossDamageFactor               = gs.BossDamageFactor,
                bossHealthFactor               = gs.BossHealthFactor,
                threatBonus                    = gs.ThreatBonus,
                pacifyAllEnemies               = gs.PacifyAllEnemies,
                tamingStartleRepercussion      = gs.TamingStartleRepercussion,
                dayTimeDuration                = (long)(gs.DayLengthMinutes  * 60_000_000_000L),
                nightTimeDuration              = (long)(gs.NightLengthMinutes * 60_000_000_000L),
                curseModifier                  = gs.CurseModifier
            };

            // Serialize user groups using camelCase
            var userGroupsNode = JsonSerializer.SerializeToNode(Config.UserGroups, GameJsonOptions)!;

            // Start from the existing file so keys owned by the server
            // (bannedAccounts, anything added in future updates) survive our rewrite.
            var serverObj = new JsonObject();
            if (File.Exists(serverJsonPath))
            {
                try
                {
                    if (JsonNode.Parse(File.ReadAllText(serverJsonPath)) is JsonObject existing)
                        serverObj = existing;
                }
                catch { }
            }

            serverObj["name"]               = Config.ServerName;
            serverObj["saveDirectory"]      = "./savegame";
            serverObj["logDirectory"]       = "./logs";
            serverObj["ip"]                 = Config.ServerIp;
            serverObj["queryPort"]          = Config.QueryPort;
            serverObj["slotCount"]          = Config.MaxPlayers;
            serverObj["tags"]               = JsonSerializer.SerializeToNode(Config.Tags);
            serverObj["voiceChatMode"]      = Config.VoiceChatMode;
            serverObj["enableVoiceChat"]    = Config.EnableVoiceChat;
            serverObj["enableTextChat"]     = Config.EnableTextChat;
            serverObj["gameSettingsPreset"] = Config.GameSettingsPreset;
            serverObj["gameSettings"]       = JsonSerializer.SerializeToNode(gameSettingsObj);
            serverObj["userGroups"]         = userGroupsNode;

            File.WriteAllText(serverJsonPath, serverObj.ToJsonString(GameJsonOptions));
            AppLogger.Info($"Updated {serverJsonPath}");
            return true;
        }
        catch (Exception ex)
        {
            AppLogger.Error($"Failed to update server JSON: {ex.Message}");
            return false;
        }
    }

    // Same shape as the passwords the server itself generates: GroupName + random suffix
    private static string GeneratePassword(string groupName)
    {
        const string chars = "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghjkmnpqrstuvwxyz23456789";
        var suffix = new char[8];
        for (int i = 0; i < suffix.Length; i++)
            suffix[i] = chars[System.Security.Cryptography.RandomNumberGenerator.GetInt32(chars.Length)];
        return groupName + new string(suffix);
    }
}
